File: CPP/Hokeemon.cpp
//Hokeemon.cpp
//Jashandeep Singh
//11/26/2022

#include <iostream>
#include <cstdlib>
#include <ctime>
#include <string>



using namespace std;

// Specification C2 - Creature Class
class Creature {
protected:
    int hunger;
    int boredom;
    string name; 

public:
// Specification A1 - Critter Name
    Creature(const string& name = "Unnamed") : hunger(rand() % 6), boredom(rand() % 6), name(name) {}

    // Specification A2 - Copy Constructor
    Creature(const Creature& other) : hunger(other.hunger), boredom(other.boredom), name(other.name) {}

    //Specification B2 - Virtual Class Creature
    virtual void play() = 0;
    virtual void feed() = 0;

    // Specification C1 - PassTime
    virtual void passTime() {
        hunger++;
        boredom++;
    }

    // Specification C4 - Overload «
    friend ostream& operator<<(ostream& os, const Creature& creature) {
        os << "Hunger: " << creature.hunger << ", Boredom: " << creature.boredom;
        return os;
    }

    // Specification B3 - Overload Assignment Operator
    Creature& operator=(const Creature& other) {
        if (this != &other) {
            hunger = other.hunger;
            boredom = other.boredom;
            name = other.name;
        }
        return *this;
    }

    // Getters for Lambda function 
    int getHunger() const { return hunger; }
    int getBoredom() const { return boredom; }
    string getName() const { return name; }

    // Component testing method
    virtual void test() = 0;
};

// Specification A3 - Second Child Class
class AquaticHokeemon : public Creature {
public:
    AquaticHokeemon(const string& name) : Creature(name) {}

    void play() override {
        int playValue = rand() % 5 + 6;
        boredom -= playValue;
        if (boredom < 0) {
            boredom = 0;
        }
        cout << "Splashing around with " << name << ". Boredom reduced by " << playValue << "." << endl;
        passTime();
    }

    void feed() override {
        int feedValue = rand() % 4 + 2;
        hunger -= feedValue;
        if (hunger < 0) {
            hunger = 0;
        }
        cout << "Feeding fish to " << name << ". Hunger reduced by " << feedValue << "." << endl;
        passTime();
    }

    void test() override {
        cout << name << " (AquaticHokeemon) is being tested." << endl;
    }
};

// Specification B1 - Child Class
class MagicalHokeemon : public Creature {
public:
    MagicalHokeemon(const string& name) : Creature(name) {}

    void play() override {
        int playValue = rand() % 5 + 4;
        boredom -= playValue;
        if (boredom < 0) {
            boredom = 0;
        }
        cout << "Playing a magical game with " << name << ". Boredom reduced by " << playValue << "." << endl;
        passTime();
    }

    void feed() override {
        int feedValue = rand() % 4 + 3;
        hunger -= feedValue;
        if (hunger < 0) {
            hunger = 0;
        }
        cout << "Feeding magical berries to " << name << ". Hunger reduced by " << feedValue << "." << endl;
        passTime();
    }

    void test() override {
        cout << name << " (MagicalHokeemon) is being tested." << endl;
    }
    friend MagicalHokeemon operator+(const MagicalHokeemon& a, const MagicalHokeemon& b) {
        string mergedName = a.getName() + "-" + b.getName();
        int mergedHunger = (a.getHunger() + b.getHunger()) / 2;
        int mergedBoredom = (a.getBoredom() + b.getBoredom()) / 2;
        MagicalHokeemon mergedHokeemon(mergedName);
        mergedHokeemon.hunger = mergedHunger;
        mergedHokeemon.boredom = mergedBoredom;
        return mergedHokeemon;
    }
};

void programGreeting();

int main() {
    programGreeting();
    srand(time(nullptr));  // Initialize random number generator

    MagicalHokeemon myHokeemon("JigglyPuff");   // Create a MagicalHokeemon object
    myHokeemon.test();                           // Run diagnostic tests


  string hokeemon1Name, hokeemon2Name;
  cout << "Enter a name for your first Hokeemon: ";
  getline(cin, hokeemon1Name);
  MagicalHokeemon hokeemon1(hokeemon1Name);

  cout << "Enter a name for your second Hokeemon: ";
  getline(cin, hokeemon2Name);
  MagicalHokeemon hokeemon2(hokeemon2Name);

  hokeemon1.test();  // Run diagnostic tests on the first Hokeemon
  hokeemon2.test();  // Run diagnostic tests on the second Hokeemon
    // Specification C3 - Validate Input
    char choice;
    do {
        cout << "Choose an action (P: Play, F: Feed, L: Listen, M: Merge, Q: Quit): ";
        cin >> choice;

        switch (choice) {
            case 'P':
            case 'p':
                myHokeemon.play();
                break;
            case 'F':
            case 'f':
                myHokeemon.feed();
                break;
            case 'L':
            case 'l':
                cout << myHokeemon;  // Using overloaded « operator
                break;
            case 'Q':
            case 'q':
                cout << "Thank you for playing!" << endl;
                return 0;
          //Specification B4- Overload + Operator
          case 'M':
          case 'm': {
              cout << "Merging Hokeemon1 and Hokeemon2." << endl;
              MagicalHokeemon mergedHokeemon = hokeemon1 + hokeemon2;
              cout << "Merged Hokeemon created: " << mergedHokeemon.getName() << endl;
              // Display merged Hokeemon details
              cout << mergedHokeemon;
              break;
          }
            default:
                cout << "Invalid choice. Please try again.\n";
        }

        myHokeemon.passTime();
        hokeemon1.passTime();
        hokeemon2.passTime();
        // Specification A4 - Lambda Function Usage
        auto displayStatus = [](const Creature& creature) {
            cout << creature.getName() << "'s Status: ";
            if (creature.getHunger() > 10) {
                cout << "Hungry, ";
            } else {
                cout << "Not Hungry, ";
            }
            if (creature.getBoredom() > 10) {
                cout << "Bored";
            } else {
                cout << "Entertained";
            }
            cout << endl;
        };

        displayStatus(myHokeemon);
        displayStatus(hokeemon1);
        displayStatus(hokeemon2);


    } while (choice != 'Q' && choice != 'q');

    return 0;
}


void programGreeting() {
    cout << "Welcome to the Hokeemon Program.\n" << endl;
    cout << "Author: Jashandeep Singh" << endl;
    cout << "Due Date: November 26, 2023" << endl;

    auto lambdaFunction = []() {
        cout << "This is a Lambda function in the Hokeemon game!\n" << endl;
    };

    lambdaFunction();
}

File: Python/Hokeemon.py
import random

class Creature:
    def __init__(self, name="Unnamed"):
        self.hunger = random.randint(0, 5)
        self.boredom = random.randint(0, 5)
        self.name = name

    def pass_time(self):
        self.hunger += 1
        self.boredom += 1

    def play(self):
        raise NotImplementedError

    def feed(self):
        raise NotImplementedError

    def test(self):
        raise NotImplementedError

    def __str__(self):
        return f"Hunger: {self.hunger}, Boredom: {self.boredom}"

class AquaticHokeemon(Creature):
    def __init__(self, name):
        super().__init__(name)

    def play(self):
        play_value = random.randint(6, 10)
        self.boredom -= play_value
        if self.boredom < 0:
            self.boredom = 0
        print(f"Splashing around with {self.name}. Boredom reduced by {play_value}.")
        self.pass_time()

    def feed(self):
        feed_value = random.randint(2, 5)
        self.hunger -= feed_value
        if self.hunger < 0:
            self.hunger = 0
        print(f"Feeding fish to {self.name}. Hunger reduced by {feed_value}.")
        self.pass_time()

    def test(self):
        print(f"{self.name} (AquaticHokeemon) is being tested.")

class MagicalHokeemon(Creature):
    def __init__(self, name):
        super().__init__(name)

    def play(self):
        play_value = random.randint(4, 8)
        self.boredom -= play_value
        if self.boredom < 0:
            self.boredom = 0
        print(f"Playing a magical game with {self.name}. Boredom reduced by {play_value}.")
        self.pass_time()

    def feed(self):
        feed_value = random.randint(3, 6)
        self.hunger -= feed_value
        if self.hunger < 0:
            self.hunger = 0
        print(f"Feeding magical berries to {self.name}. Hunger reduced by {feed_value}.")
        self.pass_time()

    def test(self):
        print(f"{self.name} (MagicalHokeemon) is being tested.")

    @staticmethod
    def merge(a, b):
        merged_name = a.name + "-" + b.name
        merged_hunger = (a.hunger + b.hunger) // 2
        merged_boredom = (a.boredom + b.boredom) // 2
        merged_hokeemon = MagicalHokeemon(merged_name)
        merged_hokeemon.hunger = merged_hunger
        merged_hokeemon.boredom = merged_boredom
        return merged_hokeemon

def program_greeting():
    print("Welcome to the Hokeemon Program.\n")
    print("Author: Jashandeep Singh")
    print("Due Date: November 26, 2023\n")

    def lambda_function():
        print("This is a Lambda function in the Hokeemon game!\n")

    lambda_function()

def main():
    program_greeting()
    random.seed()

    my_hokeemon = MagicalHokeemon("JigglyPuff")
    my_hokeemon.test()

    hokeemon1_name = input("Enter a name for your first Hokeemon: ")
    hokeemon1 = MagicalHokeemon(hokeemon1_name)

    hokeemon2_name = input("Enter a name for your second Hokeemon: ")
    hokeemon2 = MagicalHokeemon(hokeemon2_name)

    hokeemon1.test()
    hokeemon2.test()

    while True:
        choice = input("Choose an action (P: Play, F: Feed, L: Listen, M: Merge, Q: Quit): ").lower()

        if choice == 'p':
            my_hokeemon.play()
        elif choice == 'f':
            my_hokeemon.feed()
        elif choice == 'l':
            print(my_hokeemon)
        elif choice == 'm':
            print("Merging Hokeemon1 and Hokeemon2.")
            merged_hokeemon = MagicalHokeemon.merge(hokeemon1, hokeemon2)
            print(f"Merged Hokeemon created: {merged_hokeemon.name}")
            print(merged_hokeemon)
        elif choice == 'q':
            print("Thank you for playing!")
            break
        else:
            print("Invalid choice. Please try again.")

        my_hokeemon.pass_time()
        hokeemon1.pass_time()
        hokeemon2.pass_time()

        display_status = lambda creature: (
            f"{creature.name}'s Status: {'Hungry' if creature.hunger > 10 else 'Not Hungry'}, "
            f"{'Bored' if creature.boredom > 10 else 'Entertained'}"
        )

        print(display_status(my_hokeemon))
        print(display_status(hokeemon1))
        print(display_status(hokeemon2))

main()
